src/primitives/array.js

```diff
@@ -0,0 +1,14 @@
+export default class ArrayState {
+  initialize(current = []) {
+    return current;
+  }
+  push(current = [], ...args) {
```

Comment:
Conceptually it might not make sense but this is currently what happens when you have nested values. 

For example,
```
class State {
  items = Array;
}
let { states, transitions } = Microstates.from(State);
states.items;
// => []

transitions.items.push('item');
// throw Error
```

This will happen if we don't specify an argument default because the initial value is in the states tree - not inside of the transition. We would need to make the default value available in the transition somehow.

Reply:
Could it be as simple as using passing `values`  into the transition instead of `value` from the microstate constructor?

```js
constructor(tree, value) {
  let values = map(data => append(data, { value: initialize(data, value) }), tree);
  //....
  let curriedTransitions = map(
    ({ path, transitions }) =>
      map(t => (...args) => t(lensPath(path), value, ...args), transitions),
      rawTransitions);
}
```

What if we replaced `value` with `values.collapsed`?

Reply:
This is problematic because in recursive structures the values are materialized which creates lots of unnecessary objects. We could create an alternative lens set implementation that doesn't materialize getters. That'd work around this problem - I believe.

Reply:
The values won't be materialized until the transition is materialized: lazily (`tree.collapsed`  is lazy)

`map(t => (...args) => t(lensPath(path), value, ...args)` looks very much to me like `value` is never used and therefore cannot be drilled into until the transition is actually invoked.

Reply:
Oh, I think I see what you mean... You'll also need to materialize everything required to get you up to `path`? But even so, you'll need to do that in order to return the result of the transition?

Reply:
It's a bit complicated to explain. The challenge with using `values` is the fact that states tree has getters. These getters get copied into the return value of the transition. The return value is no longer a POJO. This is going to have very strange side effects. 

If I modify the microstates code to the following,

```js
    // curry in the lens and value
    let curriedTransitions = map(
      ({ path, transitions }) =>
        map(
          t => (...args) => t(lensPath(path), map(({ value }) => value, values).collapsed, ...args),
          transitions
        ),
      rawTransitions
    );
```

The following assertion,

```js
        expect(ms.transitions.contains.contains.x.increment()).toEqual({
          x: 10,
          contains: { y: 20, contains: { x: 31, y: 25 } },
        });
```

Produces this 

<img width="1436" alt="screen shot 2017-10-19 at 10 24 28 am" src="https://user-images.githubusercontent.com/74687/31754790-b64cdf04-b4b7-11e7-8aa0-f4cb82dfb7f4.png">

I did a bit more digging and found the code that [copies properties in ramda lenses].(https://github.com/ramda/ramda/blob/v0.25.0/source/assoc.js#L26) It copies enumerable properties.

If we had our own "funcadelic" lenses then we could create an `assoc` function that ignored enumerable getters. Thoughts?

Reply:
Alternatively, I could create a function that takes an object and returns an object that only has value properties, not getters.